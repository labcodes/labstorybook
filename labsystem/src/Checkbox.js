import React from "react";
import PropTypes, { string, number, bool } from "prop-types";
import { isUndefined } from "lodash";

import Icon from "./Icon";

// Checkbox //

export default class Checkbox extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    value: PropTypes.oneOfType([string, number, bool]),
    indeterminate: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    className: PropTypes.string,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    disabled: false,
    checked: undefined,
    value: null,
    indeterminate: false,
    defaultChecked: undefined,
    className: undefined,
    onChange: () => {},
  };

  constructor(props) {
    super(props);
    const { defaultChecked, checked, id } = props;
    if (!isUndefined(defaultChecked) && !isUndefined(checked)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both checked and defaultChecked for input ${id} at the same time. We always initialize the checkbox with defaultChecked. Make sure this is the behaviour you want.`
      );
    }

    let localChecked = false;
    if (defaultChecked) {
      localChecked = defaultChecked;
    } else if (!isUndefined(checked)) {
      localChecked = checked;
    }

    this.state = {
      localChecked,
    };
  }

  componentDidUpdate(prevProps) {
    const { checked } = this.props;
    if (checked !== prevProps.checked) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState(() => ({ localChecked: checked }));
    }
  }

  checkIcon = () => {
    const { disabled, indeterminate } = this.props;
    let color = "white";
    let type = "check";

    if (indeterminate) {
      type = "minus";
    }
    if (disabled) {
      color = "mineral-40";
    }

    return <Icon type={type} color={color} size="small" />;
  };

  handleOnChange = (event) => {
    const { onChange } = this.props;
    if (!isUndefined(onChange)) {
      onChange(event);
    }

    this.setState((state) => ({ localChecked: !state.localChecked }));
  };

  render() {
    const {
      id,
      name,
      className,
      label,
      disabled,
      indeterminate,
      value,
    } = this.props;

    const { localChecked } = this.state;

    return (
      <React.Fragment>
        <input
          className={`lab-checkbox ${className || ""}`}
          type="checkbox"
          id={id}
          name={name}
          checked={localChecked}
          onChange={this.handleOnChange}
          {...(value ? { value } : undefined)}
          {...(disabled ? { disabled } : undefined)}
        />
        <label className="lab-checkbox__label" htmlFor={id}>
          <span className="lab-checkbox__box">
            {localChecked || indeterminate ? this.checkIcon() : ""}
          </span>
          {label}
        </label>
      </React.Fragment>
    );
  }
}
