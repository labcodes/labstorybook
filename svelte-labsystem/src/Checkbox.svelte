<script>
  import Icon from "./Icon.svelte";

  export let id;
  export let name;
  export let label;
  export let checked = undefined;
  export let value = undefined;
  export let disabled = undefined;
  export let indeterminate = false;

  let className = "";
  export { className as class };
</script>

<input
  class={`lab-checkbox${` ${className}` || ''}`}
  type="checkbox"
  {id}
  {name}
  {value}
  {disabled}
  bind:checked
  on:change />

<label class="lab-checkbox__label" for={id}>
  <span class="lab-checkbox__box">
    {#if indeterminate && checked}
      <Icon
        type="minus"
        size="small"
        color={disabled ? 'mineral-40' : 'white'} />
    {:else if checked}
      <Icon
        type="check"
        size="small"
        color={disabled ? 'mineral-40' : 'white'} />
    {/if}
  </span>

  {label}
</label>
