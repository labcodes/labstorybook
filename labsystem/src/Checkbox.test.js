/* eslint-disable no-console */
import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import Checkbox from "./Checkbox";

describe("Checkbox", () => {
  const originalWarn = console.warn;
  afterEach(() => {
    console.warn = originalWarn;
  });

  it("renders with base props", async () => {
    expect(
      shallow(
        <Checkbox
          name="test-checkbox-group"
          id="test-checkbox"
          label="test checkbox"
        />
      )
    ).toBeTruthy();
    const renderedComponent = renderer
      .create(
        <Checkbox
          name="test-checkbox-group"
          id="test-checkbox"
          label="test checkbox"
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("raises console.warn when passing checked and defaultChecked at the same time", async () => {
    console.warn = jest.fn();

    shallow(
      <Checkbox
        name="test-checkbox-group"
        id="test-checkbox"
        label="test checkbox"
        checked
        defaultChecked
      />
    );

    const consoleText =
      "You are setting both checked and defaultChecked for input test-checkbox at the same time. We always initialize the checkbox with defaultChecked. Make sure this is the behaviour you want.";

    expect(console.warn).toBeCalledWith(consoleText);
  });

  it("renders as expected when passing disabled as true", async () => {
    const renderedComponent = renderer
      .create(
        <Checkbox
          name="test-checkbox"
          id="test-checkbox"
          label="test checkbox"
          disabled
        />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing indeterminate as true", async () => {
    const renderedComponent = renderer
      .create(
        <Checkbox
          name="test-checkbox"
          id="test-checkbox"
          label="test checkbox"
          indeterminate
          checked
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing a different className", async () => {
    const renderedComponent = renderer
      .create(
        <Checkbox
          name="test-checkbox"
          id="test-checkbox"
          label="test checkbox"
          className="lab-checkbox--purple"
          checked
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();and
  });

  it("inits state.localChecked with defaultChecked if defined", async () => {
    let component = shallow(
      <Checkbox
        name="test-checkbox"
        id="test-checkbox"
        label="test checkbox"
        defaultChecked
      />
    );
    expect(component.state().localChecked).toBe(true);

    component = shallow(
      <Checkbox
        name="test-checkbox"
        id="test-checkbox"
        label="test checkbox"
        defaultChecked={false}
      />
    );
    expect(component.state().localChecked).toBe(false);

    component = shallow(
      <Checkbox name="test-checkbox" id="test-checkbox" label="test checkbox" />
    );
    expect(component.state().localChecked).toBe(false);
  });

  it("changes state when input changes", async () => {
    const component = shallow(
      <Checkbox name="test-checkbox" id="test-checkbox" label="test checkbox" />
    );

    expect(component.state().localChecked).toBe(false);
    component.find("input").at(0).simulate("change");
    expect(component.state().localChecked).toBe(true);
  });

  it("calls props.onChange passing event when input changes", async () => {
    const mockOnChange = jest.fn();
    const component = shallow(
      <Checkbox
        name="test-checkbox"
        id="test-checkbox"
        label="test checkbox"
        onChange={mockOnChange}
      />
    );

    expect(component.state().localChecked).toBe(false);
    expect(mockOnChange).not.toBeCalled();

    component.find("input").at(0).simulate("change", { test: "event" });

    expect(component.state().localChecked).toBe(true);
    expect(mockOnChange).toBeCalledWith({ test: "event" });
  });
});
