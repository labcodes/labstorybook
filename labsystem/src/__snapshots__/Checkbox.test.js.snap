// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders as expected when passing a different className 1`] = `
<div>
  <input
    checked={true}
    className="lab-checkbox lab-checkbox--purple"
    disabled={false}
    id="test-checkbox"
    name="test-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="lab-checkbox__label"
    htmlFor="test-checkbox"
  >
    <span
      className="lab-checkbox__box"
    >
      <span
        className="lab-icon lab-icon--check lab-icon--white lab-icon--small"
      />
    </span>
    test checkbox
  </label>
</div>
`;

exports[`Checkbox renders as expected when passing disabled as true 1`] = `
<div>
  <input
    checked={false}
    className="lab-checkbox "
    disabled={true}
    id="test-checkbox"
    name="test-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="lab-checkbox__label"
    htmlFor="test-checkbox"
  >
    <span
      className="lab-checkbox__box"
    >
      
    </span>
    test checkbox
  </label>
</div>
`;

exports[`Checkbox renders as expected when passing indeterminate as true 1`] = `
<div>
  <input
    checked={true}
    className="lab-checkbox "
    disabled={false}
    id="test-checkbox"
    name="test-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="lab-checkbox__label"
    htmlFor="test-checkbox"
  >
    <span
      className="lab-checkbox__box"
    >
      <span
        className="lab-icon lab-icon--minus lab-icon--white lab-icon--small"
      />
    </span>
    test checkbox
  </label>
</div>
`;

exports[`Checkbox renders with base props 1`] = `
<div>
  <input
    checked={false}
    className="lab-checkbox "
    disabled={false}
    id="test-checkbox"
    name="test-checkbox-group"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="lab-checkbox__label"
    htmlFor="test-checkbox"
  >
    <span
      className="lab-checkbox__box"
    >
      
    </span>
    test checkbox
  </label>
</div>
`;
